syntax = "proto3";
package minecpp.proto.service.storage;

import "minecpp/proto/common/Common.proto";
import "minecpp/proto/chunk/Chunk.proto";

message ClientId {
  uint64 value = 1;
}

message Request {
  oneof message {
    RequestChunkSubscription chunk_subscription = 1;
    RequestChunkData chunk_data = 2;
    RequestSetClientId set_client_id = 3;
  }
};

message RequestChunkSubscription {
  common.ChunkPosition position = 1;
};

message RequestChunkData {
  chunk.Chunk chunk_data = 1;
};

message RequestSetClientId {
  ClientId client_id = 1;
};

message Response {
  oneof message {
    ResponseChunkData chunk_data = 1;
    ResponseEmptyChunk empty_chunk = 2;
  }
};

message ResponseEmptyChunk {
  common.ChunkPosition position = 1;
};

message ResponseChunkData {
  chunk.Chunk chunk_data = 1;
};

message ChunkSubscription {
  repeated ClientId client_ids = 1;
};